WordPress/gutenberg
# Adding support for defined IDs in `TextControl` component

## What?
This PR adds support to the `TextControl` component to accept an optional `id` prop. It currently ignores them, and generates its own.

## Why?
As per #24749, it is sometimes useful to provide components with a defined ID (for example, if you need to reference the control with an `aria-*` attribute elsewhere). Ideally all components would work this way (but that is outside the scope of this patch).

## How?
Currently, [`TextControl`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/text-control) uses [`useInstanceId`](https://github.com/WordPress/gutenberg/tree/trunk/packages/compose/src/hooks/use-instance-id) to generate an ID. As with other components (see [`SelectControl`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/select-control/index.tsx#L58), for example), this is now only used if an `id` prop is not passed in. This retains backwards compatibility, and for most use-cases will have minimal impact; IDs will continue to be generated and applied.

## Testing Instructions
Unit tests have been added to verify the expected behaviour:
 - Confirms that IDs are generated and applied, is previously
 - Confirms that ID is used if provided
 - Confirms that the `<label>` receives the ID and has a correct `for` attribute

This can also be confirmed manually:
 - Create an instance of a `TextControl`, providing both `id` and `label` props
 - Check that the `id` and `for` attributes match on the generated `<input>` and `<label>` respectively
 - Additionally, clicking on the `<label>` should put focus on the `<input>`

### Testing Instructions for Keyboard
Although this does not impact first-line keyboard access, similar manual testing can be done using a screen reader:
 - Create an instance of a `TextControl`, providing both `id` and `label` props
 - Navigate to the generated `<label>` and action it (<kbd>⌃ Control</kbd> + <kbd>⌥ Option</kbd> + <kbd>Space</kbd> with VoiceOver, for example); focus should move to the `<input>`
 - Navigate to the generated `<input>` and confirm that it has a label matching the `label` prop
